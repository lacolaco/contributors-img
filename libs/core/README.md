# core

This library was generated with [Nx](https://nx.dev).

## Running unit tests

Run `ng test core` to execute the unit tests via [Jest](https://jestjs.io).
