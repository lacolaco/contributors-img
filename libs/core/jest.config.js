const presetTs = require('ts-jest/presets').defaults;

module.exports = {
  ...presetTs,
  displayName: 'core',
  preset: '../../jest.preset.js',
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
    },
  },
  coverageDirectory: '../../coverage/libs/core',
};
