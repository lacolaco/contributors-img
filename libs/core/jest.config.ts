/* eslint-disable */
const presetTs = require('ts-jest/presets').defaults;

export default {
  ...presetTs,
  displayName: 'core',
  preset: '../../jest.preset.js',
  globals: {
    'ts-jest': {
      tsconfig: '<rootDir>/tsconfig.spec.json',
    },
  },
  coverageDirectory: '../../coverage/libs/core',
};
