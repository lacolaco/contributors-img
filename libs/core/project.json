{
  "$schema": "../../node_modules/nx/schemas/project-schema.json",
  "namedInputs": {
    "default": ["{projectRoot}/**/*"]
  },
  "tags": ["lib"],
  "projectType": "library",
  "targets": {
    "test": {
      "builder": "@nrwl/jest:jest",
      "inputs": ["default"],
      "options": {
        "jestConfig": "libs/core/jest.config.ts",
        "passWithNoTests": true
      }
    },
    "lint": {
      "builder": "@nrwl/linter:eslint",
      "inputs": ["default"],
      "options": {
        "lintFilePatterns": ["libs/core/**/*.ts"]
      }
    },
    "format": {
      "builder": "@nrwl/workspace:run-commands",
      "inputs": ["default"],
      "options": {
        "command": "yarn prettier -w '.'",
        "cwd": "libs/core"
      }
    }
  }
}
