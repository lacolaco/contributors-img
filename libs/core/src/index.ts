export * from './lib/model';
export * from './lib/state/tokens';
