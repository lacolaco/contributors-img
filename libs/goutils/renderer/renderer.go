package renderer

import (
	"bytes"
	"fmt"
	"io"
	"math"

	"contrib.rocks/libs/goutils/model"
	svg "github.com/ajstarks/svgo"
)

type RendererOptions struct {
	MaxCount int
	Columns  int
	ItemSize int
}

type renderer struct {
	Options *RendererOptions
}

func NewRenderer(o *RendererOptions) *renderer {
	return &renderer{o}
}

var _ Image = &svgImage{}

type svgImage []byte

func (i svgImage) Reader() io.ReadCloser {
	return io.NopCloser(bytes.NewReader(i))
}
func (i svgImage) Size() int64 {
	return int64(len(i))
}
func (i svgImage) ContentType() string {
	return "image/svg+xml"
}
func (i svgImage) Bytes() []byte {
	return i
}

func (r *renderer) Render(data *model.RepositoryContributors) Image {
	w := bytes.NewBuffer(nil)
	r.buildSVG(w, data)
	return svgImage(w.Bytes())
}

func (r *renderer) buildSVG(w io.Writer, data *model.RepositoryContributors) {
	itemCount := len(data.Contributors)
	columns := math.Min(float64(r.Options.Columns), float64(itemCount))
	rows := math.Ceil(float64(itemCount) / float64(columns))
	gap := 4
	width := float64(r.Options.ItemSize)*columns + float64(gap)*(columns-1)
	height := float64(r.Options.ItemSize)*rows + float64(gap)*(rows-1)

	canvas := svg.New(w)
	canvas.Start(int(width), int(height), fmt.Sprintf(`viewBox="0 0 %d %d"`, int(width), int(height)))
	for i, c := range data.Contributors {
		x := (i % int(columns)) * (r.Options.ItemSize + gap)
		y := (i / int(columns)) * (r.Options.ItemSize + gap)
		// nested <svg> is not supported in svgo
		fmt.Fprintf(canvas.Writer, `<svg x="%d" y="%d" width="%d" height="%d">\n`, x, y, r.Options.ItemSize, r.Options.ItemSize)
		{
			fillId := fmt.Sprintf("fill%d", c.ID)
			canvas.Title(c.Login)
			canvas.Circle(r.Options.ItemSize/2, r.Options.ItemSize/2, r.Options.ItemSize/2, `stroke="#c0c0c0"`, `stroke-width="1"`, fmt.Sprintf(`fill="url(#%s)"`, fillId))
			canvas.Def()
			{
				canvas.Pattern(fillId, 0, 0, r.Options.ItemSize, r.Options.ItemSize, "user")
				{
					canvas.Image(0, 0, r.Options.ItemSize, r.Options.ItemSize, c.AvatarURL)
				}
				canvas.PatternEnd()
			}
			canvas.DefEnd()
		}
		canvas.End()
	}
	canvas.End()
}
