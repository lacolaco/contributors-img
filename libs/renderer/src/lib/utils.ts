import * as SVG from '@svgdotjs/svg.js';
import AbortController from 'abort-controller';
import { JSDOM } from 'jsdom';
import fetch from 'node-fetch';

// Accept window from JSDOM
declare module '@svgdotjs/svg.js' {
  // eslint-disable-next-line @typescript-eslint/no-explicit-any
  export function registerWindow(windowImpl: any, documentImpl: any): void;
}

export async function createDataURIFromURL(imageUrl: string): Promise<string> {
  try {
    const controller = new AbortController();
    setTimeout(() => {
      controller.abort();
    }, 30 * 1000);
    const resp = await fetch(imageUrl, {
      signal: controller.signal,
    });
    const contentType = resp.headers.get('content-type');
    const data = (await resp.buffer()).toString('base64');
    return `data:${contentType};base64,${data}`;
  } catch (error) {
    console.error(error);
    return '';
  }
}

export function createSvgInstance() {
  if (SVG.getWindow() == null) {
    const DOM = new JSDOM();
    SVG.registerWindow(DOM.window, DOM.window.document);
  }
  return SVG.SVG();
}
