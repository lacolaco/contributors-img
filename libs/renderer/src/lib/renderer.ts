import { Contributor, RendererOptions } from '@lib/core';
import * as rustRenderer from '@lib/renderer-rust';

export type LayoutOptions = { itemSize: number; gap: number; maxColumns: number };
export interface ContributorsImageRenderer {
  readonly render: (contributors: Contributor[]) => Promise<string>;
  readonly layout: LayoutOptions;
}

export function createRenderer(options: RendererOptions): ContributorsImageRenderer {
  const itemSize = 64;
  const gap = 4;
  const { maxColumns, maxCount } = options;

  return {
    layout: { itemSize, gap, maxColumns },
    render: async (contributors) => rustRenderer.render(contributors.slice(0, maxCount), itemSize, maxColumns, gap),
  };
}
