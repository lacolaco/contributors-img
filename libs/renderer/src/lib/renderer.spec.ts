import { Contributor } from '@lib/core';
import { createMockContributor } from '@lib/core/testing';
import { Circle } from '@svgdotjs/svg.js';
import { ContributorsImageRenderer, createContributorView, createJsRenderer, createRustRenderer } from './renderer';
import { setupSvgRenderer } from './utils';

describe('renderer', () => {
  let renderer: ContributorsImageRenderer;
  beforeEach(() => {
    setupSvgRenderer();

    renderer = createJsRenderer({ maxColumns: 12 });
  });

  describe('render()', () => {
    test('returned value is a SVG string', async () => {
      const contributors: Contributor[] = [
        createMockContributor({
          id: 1,
          login: 'login1',
          avatar_url: 'https://via.placeholder.com',
        }),
        createMockContributor({
          id: 2,
          login: 'login2',
          avatar_url: 'https://via.placeholder.com',
        }),
      ];

      const image = await renderer.render(contributors);

      expect(image).toMatchInlineSnapshot(
        `"<svg xmlns=\\"http://www.w3.org/2000/svg\\" version=\\"1.1\\" xmlns:xlink=\\"http://www.w3.org/1999/xlink\\" xmlns:svgjs=\\"http://svgjs.dev/svgjs\\" width=\\"132\\" height=\\"64\\"><svg width=\\"64\\" height=\\"64\\" x=\\"0\\" y=\\"0\\"><title>login1</title><circle r=\\"31.5\\" cx=\\"32\\" cy=\\"32\\" stroke-width=\\"1\\" stroke=\\"#c0c0c0\\" fill=\\"url(&quot;#SvgjsPattern1000&quot;)\\"></circle><defs><pattern x=\\"0\\" y=\\"0\\" width=\\"64\\" height=\\"64\\" patternUnits=\\"userSpaceOnUse\\" id=\\"SvgjsPattern1000\\"><image width=\\"64\\" height=\\"64\\" href=\\"https://via.placeholder.com\\"></image></pattern></defs></svg><svg width=\\"64\\" height=\\"64\\" x=\\"68\\" y=\\"0\\"><title>login2</title><circle r=\\"31.5\\" cx=\\"32\\" cy=\\"32\\" stroke-width=\\"1\\" stroke=\\"#c0c0c0\\" fill=\\"url(&quot;#SvgjsPattern1001&quot;)\\"></circle><defs><pattern x=\\"0\\" y=\\"0\\" width=\\"64\\" height=\\"64\\" patternUnits=\\"userSpaceOnUse\\" id=\\"SvgjsPattern1001\\"><image width=\\"64\\" height=\\"64\\" href=\\"https://via.placeholder.com\\"></image></pattern></defs></svg></svg>"`,
      );
    });

    test('all given contributors are rendered', async () => {
      const contributors: Contributor[] = [
        { id: 1, login: 'login1', avatar_url: 'https://via.placeholder.com', html_url: 'htmlUrl1', contributions: 1 },
        { id: 2, login: 'login2', avatar_url: 'https://via.placeholder.com', html_url: 'htmlUrl2', contributions: 1 },
      ];

      const image = await renderer.render(contributors);

      expect(contributors.map((c) => c.login).every((login) => image.includes(login))).toBeTruthy();
    });
  });

  describe('createContributorView', () => {
    test('the view size is 64px by default', async () => {
      const svg = await createContributorView(createMockContributor(), { size: 64 });
      expect(svg.width()).toBe(64);
      expect(svg.height()).toBe(64);
    });

    test('the view size is configurable with options', async () => {
      const svg = await createContributorView(createMockContributor(), { size: 48 });
      expect(svg.width()).toBe(48);
      expect(svg.height()).toBe(48);
    });

    test('the view contains a title with contributor name', async () => {
      const contributor = createMockContributor();
      const svg = await createContributorView(contributor, { size: 64 });
      const [title] = svg.children();
      expect(title.svg()).toBe(`<title>${contributor.login}</title>`);
    });

    test('the view shape is a circle', async () => {
      const svg = await createContributorView(createMockContributor(), { size: 64 });
      const [_, shape] = svg.children();
      expect(shape).toBeInstanceOf(Circle);
    });

    test('the shape is filled by image', async () => {
      const contributor = createMockContributor();
      const svg = await createContributorView(contributor, { size: 64 });
      const [_, shape] = svg.children();
      const fill = svg.findOne('pattern');
      expect(shape.fill).toBeDefined();
      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      expect(shape.attr('fill')).toContain(fill!.attr('id'));
    });

    test('the image source is contributor avatar', async () => {
      const contributor = createMockContributor();
      const svg = await createContributorView(contributor, { size: 64 });

      // eslint-disable-next-line @typescript-eslint/no-non-null-assertion
      expect(svg.findOne('pattern > image')!.attr('href')).toBe(contributor.avatar_url);
    });
  });
});

describe('renderer-rust', () => {
  test('returned value is a SVG string', async () => {
    const renderer = createRustRenderer({ maxColumns: 12 });
    const contributors: Contributor[] = [
      createMockContributor({
        id: 1,
        login: 'login1',
        avatar_url: 'https://via.placeholder.com',
      }),
      createMockContributor({
        id: 2,
        login: 'login2',
        avatar_url: 'https://via.placeholder.com',
      }),
    ];
    const image = await renderer.render(contributors);

    expect(image).toMatchSnapshot();
  });
});
