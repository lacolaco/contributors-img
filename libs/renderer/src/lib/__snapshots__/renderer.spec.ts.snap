// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`renderer-rust returned value is a SVG string 1`] = `
"<svg height=\\"64\\" viewBox=\\"0 0 132 64\\" width=\\"132\\" xmlns=\\"http://www.w3.org/2000/svg\\">
<svg height=\\"64\\" width=\\"64\\" x=\\"0\\" xmlns=\\"http://www.w3.org/2000/svg\\" y=\\"0\\">
<title>login1</title>
<circle cx=\\"32\\" cy=\\"32\\" fill=\\"url('#fill1')\\" r=\\"32\\" stroke=\\"#c0c0c0\\" stroke-width=\\"1\\"/>
<defs>
<pattern height=\\"64\\" id=\\"fill1\\" patternUnits=\\"userSpaceOnUse\\" width=\\"64\\" x=\\"0\\" y=\\"0\\">
<image height=\\"64\\" href=\\"https://via.placeholder.com\\" width=\\"64\\"/>
</pattern>
</defs>
</svg>
<svg height=\\"64\\" width=\\"64\\" x=\\"68\\" xmlns=\\"http://www.w3.org/2000/svg\\" y=\\"0\\">
<title>login2</title>
<circle cx=\\"32\\" cy=\\"32\\" fill=\\"url('#fill2')\\" r=\\"32\\" stroke=\\"#c0c0c0\\" stroke-width=\\"1\\"/>
<defs>
<pattern height=\\"64\\" id=\\"fill2\\" patternUnits=\\"userSpaceOnUse\\" width=\\"64\\" x=\\"0\\" y=\\"0\\">
<image height=\\"64\\" href=\\"https://via.placeholder.com\\" width=\\"64\\"/>
</pattern>
</defs>
</svg>
</svg>"
`;
